### DISCUSSION TITLE
So many (10-12) clangd processes taking up ~35-45GB of memory

### DISCUSSION THREAD

**Discussion by vincentjgoh** [joaotavora/eglot#1236] [General]
I've got eglot managing a couple really big projects. In fact, it's the same Unreal 5 project in two different perforce workspaces.

Is there any way to limit how many processes get spun up? Or the amount of memory I'm willing to devote to this? 42GB is enough that when I compile Unreal, the distribution tool has to limit the number of parallel processes it's able to make.

And perhaps relatedly (or perhaps not), sometimes when I try to jump to a symbol, everything stalls for so long, I actually kill all the clangd processes to unlock emacs again. (It's actually only one process that's doing the locking. If I just go through and kill the active clangd processes one by one, I'll eventually happen on it, but it's not worth hunting through them when there are 10 or 12.)

![2023-06-08 20_46_25-Process Explorer - Sysinternals_ www sysinternals com  PCF_jan goh](https://github.com/joaotavora/eglot/assets/3845694/97d59e6f-8563-4b8d-bb81-0d28084c467a)

**Comment by joaotavora** [comment]
I don't know what your Emacs/Eglot  configuration is, but I'm guessingthe project configuration you have is considering 10-12 subprojects somehow, when it should consider just one.

To be able to help you you must somehow comeup with a representative project that exhibits the same behaviour and share it here. 

Please read https://joaotavora.github.io/eglot/#Troubleshooting-Eglot for instructions on how to report this bug in a way I can address it.

**Reply by vincentjgoh** [reply]
Hrm, doing that is going to be hard, since the codebase is both proprietary and massive. Trying to come up with a minimal set of circumstances to reproduce this won't be easy. But maybe I can try grabbing a clean and unencumbered Unreal project to see how it reacts. 

I CAN tell you that something has changed in the last while, but it's likely on the Unreal side, since this problem started happening with their 5.1 release. At that point, eglot seemed to tell me more often that it was managing a new project when I switched to a different directory. Would there be any way to force eglot to consider (in my case) `d:/p4v/Dev` to be my project root? How are project roots set?

Here's a list of the current eglot events buffers. 

```
*EGLOT (Debug/(c++-mode c-mode)) events*           424.6k   Fundamental  (in `d:/p4v/Dev/CoolProject/Source/CoolProject/Spawning/Debug/')
*EGLOT (Weapons/(c++-mode c-mode)) events*           1.9M   Fundamental  (in `d:/p4v/Dev/CoolProject/Source/CoolProject/Weapons/')
*EGLOT (WeaponLogic/(c++-mode c-mode)) events...     1.0M   Fundamental  (in `d:/p4v/Dev/CoolProject/Source/CoolProject/AbilitySystem/Abilities/GameplayAbilities/WeaponLogic/')
*EGLOT (Utils/(c++-mode c-mode)) events*             2.2M   Fundamental  (in `d:/p4v/DevAlt/CoolProject/Source/CoolProject/Utils/')
*EGLOT (Player/(c++-mode c-mode)) events*            1.8M   Fundamental  (in `d:/p4v/Dev/CoolProject/Source/CoolProject/Pawns/Player/')
*EGLOT (Private/(c++-mode c-mode)) events*         799.6k   Fundamental  (in `d:/p4v/Dev/Engine/Plugins/Runtime/GameplayAbilities/Source/GameplayAbilities/Private/')
*EGLOT (Engine/(c++-mode c-mode)) events*            1.0M   Fundamental  (in `d:/p4v/Dev/Engine/Source/Runtime/Engine/Classes/Engine/')
*EGLOT (Public/(c++-mode c-mode)) events*          945.5k   Fundamental  (in `d:/p4v/Dev/Engine/Plugins/Runtime/GameplayAbilities/Source/GameplayAbilities/Public/')
*EGLOT (Components/(c++-mode c-mode)) events*      796.4k   Fundamental  (in `d:/p4v/Dev/CoolProject/Source/CoolProject/AbilitySystem/Components/')
*EGLOT (UObject/(c++-mode c-mode)) events*         125.0k   Fundamental  (in `d:/p4v/Dev/Engine/Source/Runtime/Core/Public/UObject/')
```

**Reply by joaotavora** [reply]
> I CAN tell you that something has changed in the last while, but it's likely on the Unreal side, since this problem started happening with their 5.1 release. At that point, eglot seemed to tell me more often that it was managing a new project when I switched to a different directory. 

It's likely that recently those leaf directories `Debug`, `Weapons`, `GameplayAbilities` etc, gained some characteristic that made `project.el` consider them separate projects?  Can it be there's now a `.git` subdirectory in each of those?

> How are project roots set?

Why, via `project.el` interfaces :-).  There are many ways, I think, so please try to consult the documentation. Eglot has minimal interference there: it uses the project that `project.el` automatically infers (or the project that the user helped `project.el` infer, which is what you want).

Unfortunately, subproject support in `project.el` was lacking for my specific use case -- at least last I tried to set something up myself.  I ended up going a very roundabout way that I don't recommend here.  @dgutov is the the `project.el` expert.

**Reply by vincentjgoh** [reply]
I just tried everything out using lsp-mode and it seems to correctly anchor in the right root directory. When I opened a file in the DevAlt workspace, it queried to ask where it should set the project root. 

There aren't any new dot files in those directories that I listed, and if I try to query the project system it doesn't show anything (like, it just says ~ is the only project directory, which isn't very useful). 

I'm really not sure where this project behaviour is coming from, but I also don't think it's worth my time to work it out--to the extent I've ever used projects at all, I've been a projectile user.

Thanks for your help, but I think I'll just switch to lsp-mode since that seems to be working, and this particular use case would likely be a lot of work for both you and me, and I don't really want to waste either of our time. :)

**Reply by joaotavora** [reply]
> and if I try to query the project system it doesn't show anything (like, it just says ~ is the only project directory, which isn't very useful).

What does it mean to "query the project system"?

> I'm really not sure where this project behaviour is coming from

If there aren't any new dotfiles or somesuch, it's certainly coming from your configuration.

> Thanks for your help, but I think I'll just switch to lsp-mode since that seems to be working, and this particular use case would likely be a lot of work for both you and me, and I don't really want to waste either of our time. :)

No problem!
